File: android/src/main/kotlin/com/klarna/inapp/sdk/flutter_klarna_inapp_sdk/webview/WebViewHandler.kt
package com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.webview

import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.core.handler.BaseMethodHandler
import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.core.webview.WebViewManager
import io.flutter.plugin.common.MethodChannel

internal object WebViewHandler : BaseMethodHandler<WebViewMethod>(WebViewMethod.Parser) {

    var webViewManager = WebViewManager()

    override fun onMethod(method: WebViewMethod, result: MethodChannel.Result) {
        when (method) {
            is WebViewMethod.Initialize -> initialize(method, result)
            is WebViewMethod.Destroy -> destroy(method, result)
            is WebViewMethod.Show -> show(method, result)
            is WebViewMethod.Hide -> hide(method, result)
            is WebViewMethod.LoadURL -> loadURL(method, result)
            is WebViewMethod.LoadJS -> loadJS(method, result)
        }
    }

    private fun initialize(method: WebViewMethod.Initialize, result: MethodChannel.Result) {
        webViewManager.initialize(result)
    }

    private fun destroy(method: WebViewMethod.Destroy, result: MethodChannel.Result) {
        webViewManager.destroy(result)
    }

    private fun show(method: WebViewMethod.Show, result: MethodChannel.Result) {
        webViewManager.show(result)
    }

    private fun hide(method: WebViewMethod.Hide, result: MethodChannel.Result) {
        webViewManager.hide(result)
    }

    private fun loadURL(method: WebViewMethod.LoadURL, result: MethodChannel.Result) {
        webViewManager.loadURL(method.url, result)
    }

    private fun loadJS(method: WebViewMethod.LoadJS, result: MethodChannel.Result) {
        webViewManager.loadJS(method.js, result)
    }

}

File: ios/Classes/WebView/WebViewHandler.swift
import Flutter
import UIKit
import WebKit

class WebViewHandler: BaseMethodHandler<WebViewMethod> {
    
    init() {
        super.init(parser: WebViewMethods.Parser())
    }
    
    override func onMethod(method: WebViewMethod, result: @escaping FlutterResult) {
        switch method {
        case is WebViewMethods.Initialize:
            initialize(method: method as! WebViewMethods.Initialize, result: result)
        case is WebViewMethods.Destroy:
            destroy(method: method as! WebViewMethods.Destroy, result: result)
        case is WebViewMethods.Show:
            show(method: method as! WebViewMethods.Show, result: result)
        case is WebViewMethods.Hide:
            hide(method: method as! WebViewMethods.Hide, result: result)
        case is WebViewMethods.LoadURL:
            loadURL(method: method as! WebViewMethods.LoadURL, result: result)
        case is WebViewMethods.LoadJS:
            loadJS(method: method as! WebViewMethods.LoadJS, result: result)
        case is WebViewMethods.AddToHybridSDK:
            addToHybridSdk(method: method as! WebViewMethods.AddToHybridSDK, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    func initialize(method: WebViewMethods.Initialize, result: @escaping FlutterResult) {
        
    }
    
    func destroy(method: WebViewMethods.Destroy, result: @escaping FlutterResult) {
        
    }
    
    func show(method: WebViewMethods.Show, result: @escaping FlutterResult) {
        
    }
    
    func hide(method: WebViewMethods.Hide, result: @escaping FlutterResult) {
        
    }
    
    func loadURL(method: WebViewMethods.LoadURL, result: @escaping FlutterResult) {
        
    }
    
    func loadJS(method: WebViewMethods.LoadJS, result: @escaping FlutterResult) {
        
    }
    
    func addToHybridSdk(method: WebViewMethods.AddToHybridSDK, result: @escaping FlutterResult) {
        
    }
    
}
